### DISCUSSION TITLE
Error 500: Internal server error with getServerSideProps on production

### DISCUSSION THREAD

**Discussion by kazeemOnis** [vercel/next.js#20641] [Help]
Hi good day, I'm having an issue on production with a nextjs project. I get a 500 error returned on each page that uses getServerSideProps to make requests to the database on my VPS, it works well with no issues locally. Each page requests get called on the server and it returns but the props data is never passed onto the page. Attached below is how I'm using getServerSideProps, build log on my VPS and custom _app.js

![image](https://user-images.githubusercontent.com/20818084/103405680-2ed75000-4b58-11eb-84f6-857fd1c627bd.png)

![image](https://user-images.githubusercontent.com/20818084/103405741-7231be80-4b58-11eb-9a07-754819833dd8.png)

![image](https://user-images.githubusercontent.com/20818084/103405808-af964c00-4b58-11eb-8d4c-73291f4a8e3c.png)

**Comment by teksebarna** [comment]
Did you find a solution for it?
I'm getting error 500 after I made the Next.js upgrade from 10.0.3 to 10.2.0

**Reply by bobdeei** [reply]
I'm using 10.2.2 and still runs into this issue. Still in the process of figuring it out.

**Reply by kavinvalli** [reply]
Did someone figure this out?

**Reply by AtoLrn** [reply]
Same Here :(

**Reply by AP-DAVID** [reply]
Same here. I am looking for a solution.

**Reply by kbnewlon** [reply]
Anyone find a solution?

**Reply by shonsirsha** [reply]
So no solutions yet?

**Reply by leedoopal** [reply]
Same Here :( I made next 12.1.6

**Comment by AlexanderKozhevin** [comment]
same problem

```
export async function getServerSideProps({ query }) {
  const pid = query.pid
  return { props: { pid }}
}
```
<img width="717" alt="Screenshot 2022-02-04 at 16 31 41" src="https://user-images.githubusercontent.com/4464561/152537397-c1482571-7935-4dc5-abc1-46737db91af1.png">

**Comment by erikmaddy** [comment]
It’s crazy that this has been up since the end of 2020. This happened to me as well. Nothing changed in my project but all of a sudden, boom! 500 error. Same exact structure as @AlexanderKozhevin. It has been working fine up until a couple days ago.

**Reply by stuvms** [reply]
Hi, i had the same problem described on top (im not an expert or anything but ive gone through this error on production), i could resolve it by upgrading the version of the following dependencies: 
    "node": "8.11.x",
    "npm": "8.19.3"
then i replace (these dependencies seems to be deprecated):
@material-ui/icons  --> @mui/icons-material
@material-ui/core --> @mui/material
as soon as a replace those, the 500 error is no longer presented, hope this helps.

**Comment by itwasmattgregg** [comment]
This is probably not the same situation but it's what was happening for me. I was sending a large payload (2.4MB) across the server side props and I was not getting a 500 locally but was in production hosted with Vercel. I was not getting any error messages on Vercel server. Totally fixed when I moved that large payload to load on the page.

**Reply by ghost** [reply]
Hey! I am experiencing this as well, could you explain what you mean by "fixed when I moved that large payload to load on the page". I am trying to fetch data from mongodb inside getServerSideProps() and followed this guide https://www.mongodb.com/developer/how-to/nextjs-with-mongodb/

I am most definitely not going with this simple solution further on, but I am curious how to fix this issue.

**Reply by itwasmattgregg** [reply]
Yea I mean I created an api route that fetches the data from the db and then I call that in a useEffect after page load. It only happens on large data queries though.

**Comment by stephencweiss** [comment]
@itwasmattgregg - your situation sounds very similar to mine. I'll definitely try out the API route to see if that fixes my issue.

I'm curious though: do we know _why_ this is happening? It seems to me that this flow isn't "wrong" - so I'm struggling to understand the cause and why the API would actually be a fix.

**Reply by itwasmattgregg** [reply]
There must be some upper limit on data loaded in getServerSideProps? That's my only guess

**Reply by stephencweiss** [reply]
Would be great if that could get documented somewhere. I'm chasing ghosts right now.

**Reply by tankez0r** [reply]
@itwasmattgregg  I got the same problem, I think the issue is exactly what you say. It radicate in  the time of the request. In my case, when I call for a response in my api (working in the same host) it works correctly, but when I try to make a call from an external api and this take his time, nextjs try to build a page quickly, not making a fallback to await the response. 
 Maybe a solutión (for now), Is create a client side rendering using SWR. I got the same problem on my host.
I think this issue is generated in the build, because it work perfectly on developing, maybe something who is virtualized in dev and no contemplated to work on the same form on production.
 Sorry if my english is poor,  I will try to use swr and then I will come to tell the results

**Comment by shonsirsha** [comment]
This has just happened to me. Honestly it's very stressful - ive spent way too much time wondering what was wrong with my API, only to find out (ish) here that it's some NextJS shenanigans. I guess we'll have to client render it then...

**Comment by tankez0r** [comment]
Hello, I found a solution. Is a version problem. In general this issues are cause by intern errors in the new libraries of js using by vercel. 
In the only place where my code is working great is in vercel servers. In the server before i deployed in vercel I was using nodejs 16.15.0. Maybe the error is in your nextjs core versión or in the nodejs version. 
I'm sorry for my english, I hope this work for you.

PD: Remember you can handle errors with trycatch. If you don't handle the error, this automatically jump to error 500. 
to follow the error stack just return ther error in the catch as object.

**Comment by icyJoseph** [comment]
Are you all perhaps fetching more than 5MB of data? I think there might be some limitation there.

**Reply by leedoopal** [reply]
thanks. but I'm fetching 1.2kb data 😢

**Reply by icyJoseph** [reply]
No problem, but that's one less thing to worry about.

Do you use env vars to fetch this data? I've experienced that sometimes these are missing.

Where/how do you deploy?

Are there any logs?

**Reply by icyJoseph** [reply]
Also, do you have the same node.js version on your server and on your local environment?

**Reply by tankez0r** [reply]
What versión of next js are you using? I fix all this problems using vercel (because an argentinian host that I use to deploy have issues with his javascript versions). The good think in vercel with this troubles is if they find an error in the deployment, the console return a battery with the issue.

**Reply by shonsirsha** [reply]
Hey, I solved mine by changing the Serverless Function Region to somewhere else other than the preselected one - and yeah it worked this way.

**Reply by leedoopal** [reply]
Thank you.
I transferred the API that was being called on the server-side to the client.
Then I modified fetch data that was being called using env var. 
I solved it :)

nextjs v12.2

**Comment by sharonjmoses** [comment]
return {
props : 
{
data
}
}

in this case data will be considered as null, Since it is null server throwing 500 error.  Return only when you get the data, this will fix the issue

**Reply by bobdeei** [reply]
It's too long since I had this problem that I don't even remember what project I was on. But I'll try yours next time. Thanks @sharonjmoses

**Comment by atco-jh** [comment]
I opened an issue regarding this: #47642 

I don't think it's an upper limit issue, unless the upper limit is incredibly small. I was able to replicate the error with a very minimal example passing only a string through the data

**Comment by thiagomurtinho** [comment]
The getServerSideProps function is a fragile feature of the Next.js framework and should be used with care to avoid server crashes. It is highly recommended to have adequate error handling when implementing this function. A case study illustrates this situation: in a project, the getStaticProps function was replaced by getServerSideProps, which caused problems. While Next.js ignored updates to static pages when using getStaticProps with the revalidate parameter, the same did not happen with getServerSideProps.

The solution was to implement a more robust error handling within the getServerSideProps function, especially when making calls to the backend.

Basic fetcher:
```(ts)
const fetcher = (url: string) => fetch(url).then(r => r.json())  //  Shared usage with SWR is the reason for this format.
```

Fixed:
```(ts)
export const fetcher = async (url: string) => {
  try {
    const response = await fetch(url)

    if (!response.ok) {
      throw new Error(`Failed to retrieve data ${url}}`)
    }

    const data = await response.json()

    return data
  } catch (error) {
    console.error(error)
  }
}
```

getServerSideProps fixed:
```(ts)
export const getServerSideProps: GetServerSideProps = async ctx => {
 // ... implementations
  const fetcherRes = await fetcher (`${process.env.API_URL}/api/endpoint`)

  return {
    props: {
      fetcherRes ,
    // ...
    }
  }
}
```

It's just an example and it's worth remembering that I advise not to use third-party axios or libs.

**Reply by atco-jh** [reply]
Hey, thanks for the response, but I'm still unsure how this is causing a server crash? for example, I made a very simple project that uses `getServerSideProps` 

pages/posts/[id].tsx
``` 
import Posts from "@/lib/components/Posts";

export default function Page(props: any) {
  return <Posts apiData={props.apiData} />;
}

export async function getServerSideProps({ params }: any) {
  return {
    props: {
      success: true,
      apiData: JSON.parse(JSON.stringify({ id: params.id })),
    },
  };
}
```

Here's how it's linked to in other pages: 
```
   router.push({
      pathname: "/posts/[id]",
      query: { id: name },
    });
```

Firebase function to serve the pages: 
```
const { https } = require("firebase-functions");
const { default: next } = require("next");

const nextjsServer = next({
  dev: false,
  conf: { distDir: "../../next" },
});
const nextjsHandle = nextjsServer.getRequestHandler();

exports.nextjsFunc = https.onRequest((req, res) => {
  return nextjsServer.prepare().then(() => nextjsHandle(req, res));
});
```

Full example repo [here](https://github.com/atco-jh/nextjs-firebase-ssg-ssr-example)

So in this instance, it's not even calling an outside API, and there is no library. It works perfectly locally, and on an emulated firebase server, but not when hosted. What am I doing wrong?

**Reply by thiagomurtinho** [reply]
I'm not sure if I understood correctly the reason for the format of your implementation, but to retrieve the ID in /posts/[id], I would use another method that is more recommended in the documentation.

It would look something like this.
pages/posts/[id].tsx
```
import { useRouter } from 'next/router'
import Posts from "@/lib/components/Posts";

export default function Page(props: any) {
  const router = useRouter()
  const { id } = router.query
  
  return <Posts apiData={id} />;
}
```

The Firebase function doesn't seem to be wrong.

**Reply by eduardourbina-lineup** [reply]
Hello @thiagomurtinho, I'm currently using the app directory. This won't allow me to do that like this, but the useParams() and that's the hook I'm using to get the id, but I'd still get the internal server error. 

```
  const { id } = router.query; //Property 'query' does not exist on type 'AppRouterInstance'.ts(2339)
```

When you do a dynamic route, the fetching thorugh SSR is mandatory?

**Comment by vikingsurfeur** [comment]
Thank you @thiagomurtinho , I implement your suggestion and it works. I give one more thing in my fetcher function, we should abort the request if the response is not ok, we can do that with the AbortController:

```
import { baseUrlApi, envResolver } from "./envResolver"

export const fetcher = async (resource: string, queryParams?: string) => {
    try {
        let fetchUrl;

        queryParams
            ? fetchUrl = `${baseUrlApi}/${resource}?${queryParams}`
            : fetchUrl = `${baseUrlApi}/${resource}`;

        const controller = new AbortController();
        const { signal } = controller;

        const res = await fetch(fetchUrl, {
            headers: {
                Authorization: `Bearer ${envResolver.apiKey}`,
            },
            signal,
        });

        if (!res.ok) {
            controller.abort;
            throw new Error(`Failed to retrieve data / Error: ${res.status}`);
        }

        const data = await res.json();

        return data;
    } catch (error) {
        console.error(error);
    }
}
```
Take care Thia, greetings from France

**Comment by ATPAustinPeng** [comment]
I'm pretty sure I have the exact same issue as the original thread (api works in local env; 500 internal error in deployment) and I fixed it by using `mongodb` instead of `mongoose` to create my DB connection. I literally clone this [Official MongoDB Reference repository](https://github.com/mongodb-developer/nextjs-with-mongodb) and tested it with my MongoDB URI and data collection and it worked. I hope it helps. @kazeemOnis @thiagomurtinho

**Reply by pablojsx** [reply]
I'm not even using a DB yet, just  simple fetch from an API and it won't work:

`export async function getServerSideProps(ctx: GetServerSidePropsContext) {
  try {
    const menu = await getMenu();
    return { props: { menu } };
  } catch {
    return { props: { menu: [], error: { message: "Couldn't get the menu." } } };
  }
}`

**Comment by snehilgamit** [comment]
if you are using nextAuth then check you nextAuth_url in .env or check your server logs.
Most web hosting providers and server environments provide logs that can give you more detailed information about what went wrong. Look for any error messages or stack traces in these logs.

**Comment by TianMeh** [comment]
Hi if this helps anyone I've solved it by downgrading nextjs from 13.5.4 to 13.5.3
Before I had all error handling and everything but still did not work. Downgrading the version solved it for me

https://www.reddit.com/r/nextjs/comments/170shch/works_on_localhost_but_not_on_vercel/
